### PR TITLE
Contact Diary - Status Quo (Conclusion-Draft) (EXPOSUREAPP-3034)

### PR DESCRIPTION
**Pull request by d4rken**
This PR (currently draft) reflects the current development of the contact-diary.
Corresponding Ticket: EXPOSUREAPP-3034
All further work will be squash-merged into this PR (Overview).

### CODE REVIEW

**Review by ralfgehrer — APPROVED** [review]
I tested all common functionalities of the feature and it looks pretty good. Some UI polishments are outstanding - such as the button heights (too small) or text colors of disabled buttons. Will be fixed in upcoming smaller PRs.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/retention/ContactDiaryCleanTask.kt [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        Timber.i("Finished (isCanceled=$isCanceled).")
+        internalProgress.close()
+    }
+
+    override suspend fun cancel() {
+        Timber.w("cancel() called.")
+        isCanceled = true
+    }
+
+    class Config : TaskFactory.Config {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/retention/ContactDiaryCleanTask.kt:46]: Forgot to set error handling to SILENT

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/ContactDiaryStorageModule.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import de.rki.coronawarnapp.contactdiary.storage.dao.ContactDiaryLocationDao
+import de.rki.coronawarnapp.contactdiary.storage.dao.ContactDiaryLocationVisitDao
+import de.rki.coronawarnapp.contactdiary.storage.dao.ContactDiaryPersonDao
+import de.rki.coronawarnapp.contactdiary.storage.dao.ContactDiaryPersonEncounterDao
+import de.rki.coronawarnapp.contactdiary.storage.repo.ContactDiaryRepository
+import de.rki.coronawarnapp.contactdiary.storage.repo.DefaultContactDiaryRepository
+import javax.inject.Singleton
+
+@Module
+class ContactDiaryStorageModule {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/ContactDiaryStorageModule.kt:14]: Could cause the main thread to hang if this causes the whole DB to be instantiated on DI graph instantiation.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/repo/ContactDiaryRepository.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    suspend fun updateLocation(contactDiaryLocation: ContactDiaryLocation)
+    suspend fun deleteLocation(contactDiaryLocation: ContactDiaryLocation)
+    suspend fun deleteLocations(contactDiaryLocations: List<ContactDiaryLocation>)
+    suspend fun deleteAllLocations()
+
+    // Location visit
+    val locationVisits: Flow<List<ContactDiaryLocationVisit>>
+    fun locationVisitsForDate(date: LocalDate): Flow<List<ContactDiaryLocationVisit>>
+    suspend fun addLocationVisit(contactDiaryLocationVisit: ContactDiaryLocationVisit)
+    suspend fun deleteLocationVisit(contactDiaryLocationVisit: ContactDiaryLocationVisit)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/repo/ContactDiaryRepository.kt:25]: A single deletion method that accepts a list is sufficient.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/ContactDiaryActivity.kt [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import dagger.android.HasAndroidInjector
+import de.rki.coronawarnapp.R
+import de.rki.coronawarnapp.util.di.AppInjector
+import javax.inject.Inject
+
+/**
+ * This activity holds all the contact diary fragments
+ */
+class ContactDiaryActivity : AppCompatActivity(), HasAndroidInjector {
+    companion object {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/ContactDiaryActivity.kt:21]: Companion objects the end of the class.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/ContactDiarySettings.kt [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import android.content.Context
+import de.rki.coronawarnapp.util.di.AppContext
+import de.rki.coronawarnapp.util.preferences.clearAndNotify
+import de.rki.coronawarnapp.util.preferences.createFlowPreference
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class ContactDiarySettings @Inject constructor(
+
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/ContactDiarySettings.kt:12]: Weird formatting

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/ContactDiaryUIModule.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import de.rki.coronawarnapp.contactdiary.ui.day.tabs.person.ContactDiaryPersonListFragment
+import de.rki.coronawarnapp.contactdiary.ui.day.tabs.person.ContactDiaryPersonListModule
+import de.rki.coronawarnapp.contactdiary.ui.edit.ContactDiaryEditModule
+import de.rki.coronawarnapp.contactdiary.ui.onboarding.ContactDiaryOnboardingFragment
+import de.rki.coronawarnapp.contactdiary.ui.onboarding.ContactDiaryOnboardingFragmentModule
+import de.rki.coronawarnapp.contactdiary.ui.overview.ContactDiaryOverviewFragment
+import de.rki.coronawarnapp.contactdiary.ui.overview.ContactDiaryOverviewFragmentModule
+
+@Module
+    (
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/ContactDiaryUIModule.kt:22]: Weird formatting

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditLocationsFragment.kt [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    @Inject lateinit var viewModelFactory: CWAViewModelFactoryProvider.Factory
+    private val viewModel: ContactDiaryEditLocationsViewModel by cwaViewModels { viewModelFactory }
+    private val binding: ContactDiaryEditLocationsFragmentBinding by viewBindingLazy()
+
+    private val locationList: MutableList<ContactDiaryLocation> = mutableListOf()
+    private val listAdapter = ListAdapter()
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        binding.viewModel = viewModel
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditLocationsFragment.kt:37]: Please don't assign viewmodels directly. Due to the spaghetti code it creates we don't want databinding to do direct access of LiveData. Pass state objects.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditLocationsFragment.kt [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            R.string.contact_diary_delete_button_negative,
+            positiveButtonFunction = {
+                viewModel.onDeleteAllConfirmedClick()
+            }
+        )
+    }
+
+    inner class ListAdapter : RecyclerView.Adapter<ListAdapter.ViewHolder>() {
+
+        inner class ViewHolder(listItemView: View) : RecyclerView.ViewHolder(listItemView) {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditLocationsFragment.kt:91]: Why is this an inner class?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditPersonsViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModel
+import de.rki.coronawarnapp.util.viewmodel.SimpleCWAViewModelFactory
+import kotlinx.coroutines.flow.map
+
+class ContactDiaryEditPersonsViewModel @AssistedInject constructor(
+    private val contactDiaryRepository: ContactDiaryRepository,
+    dispatcherProvider: DispatcherProvider
+) : CWAViewModel(dispatcherProvider = dispatcherProvider) {
+
+    val personsLiveData = contactDiaryRepository.people.asLiveData()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditPersonsViewModel.kt:20]: I think the database access can happen on the main thread here.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/overview/ContactDiaryOverviewFragment.kt [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        binding.apply {
+            contactDiaryOverviewRecyclerview.adapter = adapter
+
+            contactDiaryOverviewHeader.contactDiaryHeaderButtonBack.buttonIcon.setOnClickListener {
+                vm.onBackButtonPress()
+            }
+        }
+
+        vm.listItems.observe2(this) {
+            adapter.setItems(it)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/overview/ContactDiaryOverviewFragment.kt:43]: Does this not need `notifyDataSetChanged`?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt [hunk 11]
```diff
@@ -90,6 +90,9 @@ class HomeFragment : Fragment(R.layout.fragment_home), AutoInject {
                 HomeFragmentEvents.ShowDeleteTestDialog -> {
                     showRemoveTestDialog()
                 }
+                HomeFragmentEvents.GoToContactDiary -> {
+                    (requireActivity() as MainActivity).navigateToContactDiaryActivity()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt:94]: Why does this go through the activity? Are there are navigation paths besides HomeFragment?

**Review by d4rken — COMMENTED** [review]
Small stuff, no blockers, can be fixd in a different PR.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditPersonsViewModel.kt (hunk 9 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditPersonsViewModel.kt:20]: Same for the other repo access calls below in this file.

**Review by BMItr — APPROVED** [review]
Device test succesul (Pixel 2).
Spotted smaller cleancode issues. 
Overall best MVP I have seen so far :heart:

**Review by axelherbstreith — APPROVED** [review]

**Comment by mlenkeit** [comment]
Installing and testing 8e63b803 this morning to see if I find anything major...

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/ContactDiaryActivity.kt (hunk 4 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/ContactDiaryActivity.kt:21]: https://github.com/corona-warn-app/cwa-app-android/pull/1943
adjusted with upcoming codesmell PR

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/ContactDiarySettings.kt (hunk 5 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/ContactDiarySettings.kt:12]: cleaned with https://github.com/corona-warn-app/cwa-app-android/pull/1943.
upcoming.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/retention/ContactDiaryCleanTask.kt (hunk 1 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/retention/ContactDiaryCleanTask.kt:46]: https://github.com/corona-warn-app/cwa-app-android/pull/1936/files

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/overview/ContactDiaryOverviewFragment.kt (hunk 10 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/overview/ContactDiaryOverviewFragment.kt:43]: yes.
its already handled inside of setItems(it)
_fun setItems(elements: List<ListItem>) {
        this.elements.clear()
        this.elements += elements
        notifyDataSetChanged()
    }_

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditLocationsFragment.kt (hunk 7 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditLocationsFragment.kt:37]: What spaghetti code do you mean? I think it's much cleaner this way.

Comment (AlexanderAlferov) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditLocationsFragment.kt:37]: Addressed in #1951

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditPersonsViewModel.kt (hunk 9 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditPersonsViewModel.kt:20]: Is there a benefit to having it on the main thread?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/ContactDiaryStorageModule.kt (hunk 2 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/ContactDiaryStorageModule.kt:14]: What is your suggestion? I haven't noticed a problem.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/repo/ContactDiaryRepository.kt (hunk 3 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/repo/ContactDiaryRepository.kt:25]: This method is being used and works fine. Why refactor?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/ContactDiaryStorageModule.kt (hunk 2 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/ContactDiaryStorageModule.kt:14]: I'm already addressing this...

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt (hunk 11 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt:94]: https://github.com/corona-warn-app/cwa-app-android/pull/1952

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditPersonsViewModel.kt (hunk 9 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditPersonsViewModel.kt:20]: I worded it incorrectly, sorry, I meant: We should make sure to not access the database on the main thread.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/ContactDiaryUIModule.kt (hunk 6 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/ContactDiaryUIModule.kt:22]: adjusted.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/ContactDiaryStorageModule.kt (hunk 2 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/ContactDiaryStorageModule.kt:14]: done.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditPersonsViewModel.kt (hunk 9 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditPersonsViewModel.kt:20]: Ok. I thought you were asking to have it on the main thread.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/repo/ContactDiaryRepository.kt (hunk 3 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/storage/repo/ContactDiaryRepository.kt:25]: You both are right.
it is not really needed but its like a comfort feature that you not have to create a list for adding a simple single item.
Would let it as it is for now.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditPersonsViewModel.kt (hunk 9 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/contactdiary/ui/edit/ContactDiaryEditPersonsViewModel.kt:20]: https://github.com/corona-warn-app/cwa-app-android/pull/1954/files

**Pull request merged**